core/src/main/scala/kafka/server/DelayedOperation.scala

```diff
@@ -101,16 +102,35 @@ abstract class DelayedOperation(override val delayMs: Long,
   /**
    * Thread-safe variant of tryComplete() that attempts completion only if the lock can be acquired
    * without blocking.
+   *
+   * If threadA acquires the lock and performs the check for completion before completion criteria is met
+   * and threadB satisfies the completion criteria, but fails to acquire the lock because threadA has not
+   * yet released the lock, we need to ensure that completion is attempted again without blocking threadA
+   * or threadB. `tryCompletePending` is set by threadB when it fails to acquire the lock and at least one
+   * of threadA or threadB will attempt completion of the operation if this flag is set. This ensures that
+   * every invocation of `maybeTryComplete` is followed by at least one invocation of `tryComplete` until
+   * the operation is actually completed.
    */
   private[server] def maybeTryComplete(): Boolean = {
+    val done = tryCompleteIfLockIsFree()
```

Comment:
After thinking some more, I'm not sure this is 100% bulletproof. I think it works for two concurrent calls, but consider this sequence:

1. Thread A is in `tryComplete`.
2. Thread B calls `maybeTryComplete` and sets `tryCompletePending` to true.
3. Thread B exits `maybeTryComplete`.
4. Thread A finishes `tryComplete` and observes `tryCompletePending` as true
5. Thread A goes back into `tryComplete`
6. Thread B calls `maybeTryComplete` and sets `tryCompletePending` to true.
7. Thread B exits `maybeTryComplete`.
8. Thread A finishes `tryComplete` and exits `maybeTryComplete` with `tryCompletePending` set to true.

To fix it, I think we somehow need a loop to ensure that we retry whenever the holder of the lock observes `tryCompletePending` as true following the call to `tryComplete`. Does this work?

```
 private def tryCompleteIfLockIsFree(): Boolean = {
    if (lock.tryLock()) {
      try {
        do {
          tryCompletePending.set(false)
          tryComplete()
        } while (tryCompletePending.get())
      } finally {
        lock.unlock()
      }
    } else {
      tryCompletePending.set(true)
      false
    }
  }
```

Reply:
@hachikuji I started with a loop yesterday and somehow convinced myself that a loop was not required. But you are right, we do need a loop and your code looks good. I will update the PR and also try to include your scenario in the unit test. Thank you!

Reply:
@hachikuji I have updated the code, but haven't managed to write a test for the scenario yet. It is turning out to be a bit harder to hit this window in a test.

Reply:
@hachikuji Not sure if I missed something, but I think we need to check `tryCompletePending` after we release the lock. Otherwise, ThreadA might just have finished the loop inside the lock when ThreadB sets the flag. I pushed another commit with a slightly different loop. Haven't yet figured out a way to test this, will work on that later today. If you can think of any obvious holes in the code, please let me know.

Reply:
Yes, I think you are right. I was about to send a comment saying the same thing 😉 .